authz/policy/common/src/main/java/org/keycloak/authorization/policy/provider/js/JSPolicyProviderFactory.java

```diff
@@ -19,9 +24,11 @@
  */
 public class JSPolicyProviderFactory implements PolicyProviderFactory<JSPolicyRepresentation> {
 
-    private static final String ENGINE = "nashorn";
+    private static String ENGINE = "nashorn";
 
-    private JSPolicyProvider provider = new JSPolicyProvider(() -> new ScriptEngineManager().getEngineByName(ENGINE));
+    private JSPolicyProvider provider = new JSPolicyProvider(this::getCompiledCode);
+    private Compilable engine;
+    private Map<String, CompiledScript> compiled = Collections.synchronizedMap(new HashMap<>());
```

Comment:
We can probably use a simple LRU map to cache entries. Where the number of cached entries can be defined by users.

There is a simple LRU map implementation we are already using in other places, we can probably re-use the same code here.